### PR TITLE
[BugFix] The short key index maybe not order by sort key columns

### PR DESCRIPTION
**Pull request by sevev**
This bug is introduced by this pr(https://github.com/StarRocks/starrocks/pull/25286), we fix the bug which short key index and segment data may not be consistent after vertical compaction in the pr. But the fix ignores the following scenario:
if we create a table which order by column `c2` and c1 and `c2` is after `c1`,
```
CREATE TABLE pk1sk2_tbl (
    k1 int,
    c1 varchar(20),
    c2 varchar(20)
PRIMARY KEY(k1)
DISTRIBUTED BY HASH(c1) BUCKETS 3
ORDER BY(c2,c1);
```
The data of column `c2` should be the prefix of short key index, but the following code in `segment_writer.cpp`:
```
    for (uint32_t i = 0; i < _column_indexes.size(); i++) {
        uint32_t column_index = _column_indexes[i];
        ....
        if (column.is_sort_key()) {
            _sort_column_indexes.emplace_back(i);
        }
    }
```
When we import data or do horizontal compaction, segment_writer's init schema is tablet schema, and the `_column_indexes` is ordered by table column. At this time, column `c1` and column `c2` are both sort key columns and we check column `c1` first, so the `_sort_column_indexes` keeps the index of `c1` first. And in the following short key index encoding process, the data of `c1` will be written before `c2` and be the prefix of the short key index.

The solution is to check `_sort_column_indexes` and the elements of `_sort_column_indexes` should be ordered by the sort key column index.

## What type of PR is this:
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [x] 3.0
  - [x] 2.5
  - [ ] 2.4

### CODE REVIEW

File: be/src/storage/rowset/segment_writer.cpp [hunk 1]
```diff
@@ -181,7 +182,21 @@ Status SegmentWriter::init(const std::vector<uint32_t>& column_indexes, bool has
...
+    for (auto& column_idx : _tablet_schema->sort_key_idxes()) {
+        auto iter = sort_column_idx_by_column_index.find(column_idx);
+        if (iter != sort_column_idx_by_column_index.end()) {
+            _sort_column_indexes.emplace_back(iter->second);
+        } else {
+            // Currently we have the following two scenarios：
+            //  1. data load or horizontal compaction, we will write the whole row data once a time
+            //  2. vertical compaction, we will first write all sort key columns and write value columns by group
+            // So the all sort key columns should be found in `_column_indexes` so far.
+            std::string err_msg =
```
Comment (chaoyli) [inline be/src/storage/rowset/segment_writer.cpp:197]: Add the `tablet_id` into the error log.

**Review by chaoyli — DISMISSED** [review]

**Review by decster — DISMISSED** [review]

**Review by decster — APPROVED** [review]

**Review by Astralidea — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.5

**Comment by sevev** [comment]
@mergify backport branch-3.0

**Pull request merged**